File: app/src/main/java/org/silvermoon/osm_kotlin/concurrency/RemoteAsyncTileLoader.java
package org.silvermoon.osm_kotlin.concurrency;

import android.os.Handler;
import android.os.Message;

import org.silvermoon.osm_kotlin.mapunits.OsmMapViewBase;
import org.silvermoon.osm_kotlin.mapunits.Tile;
import org.silvermoon.osm_kotlin.model.entities.MapTile;

public class RemoteAsyncTileLoader {

    private static final int NB_THREAD_FOR_TILE_REQUEST = 3;

    private TileRequestTask[] mDownloadTileTasks;
    private int mIterator;


    public RemoteAsyncTileLoader(Handler handler) {

        mIterator = 0;
        mDownloadTileTasks = new TileRequestTask[NB_THREAD_FOR_TILE_REQUEST];

        for (int i = 0; i < mDownloadTileTasks.length; i++) {
            mDownloadTileTasks[i] = new TileRequestTask(handler);
        }
    }

    public void queueTileRequest(Tile tile) {

        // Make sure the zoom is not > to the actual max Tile OSM zoom 
        if (tile.getZoom() > OsmMapViewBase.MIN_ZOOM_LEVEL_FOR_TILES)
            return;
        //---

        // Check if tile is not already in a thread
        for (int i = 0; i < mDownloadTileTasks.length; i++) {
            if (mDownloadTileTasks[i].mRequestsQueue.contains(tile))
                return;
        }
        //---

        // if not add it
        mDownloadTileTasks[mIterator].queue(tile);
        //--

        mIterator++;
        if (mIterator == mDownloadTileTasks.length)
            mIterator = 0;
    }

    public void interruptThreads() {
        for (int i = 0; i < mDownloadTileTasks.length; i++) {
            mDownloadTileTasks[i].interrupt();
        }
    }

    private class TileRequestTask extends Thread {

        private static final int MAX_QUEUE_REQUEST = 6;

        public RequestsQueue mRequestsQueue;

        private RequestTile mRequestTile;
        private Handler mHandler;

        public TileRequestTask(Handler handler) {
            mHandler = handler;
            mRequestTile = new RequestTile();
            mRequestsQueue = new RequestsQueue(1, MAX_QUEUE_REQUEST);
            start();
        }

        public void queue(Tile tile) {
            mRequestsQueue.queue(tile);

            synchronized (this) {
                this.notify();
            }
        }

        @Override
        public void run() {

            Tile tile;

            while (true) {
                tile = null;
                if (mRequestsQueue.hasRequest()) {
                    tile = mRequestsQueue.dequeue();
                }
                if (tile != null) {
                    loadTileAndSendMessage(tile);
                }
                try {
                    synchronized (this) {
                        if (mRequestsQueue.size() == 0) {
                            this.wait();
                        }
                    }
                    Thread.sleep(150);
                } catch (InterruptedException e) {
                    break;
                }
            }
        }

        public void loadTileAndSendMessage(Tile remoteTile) {

            if (remoteTile != null && remoteTile.getKey() != null) {
                try {
                    remoteTile.setBitmap(mRequestTile.loadBitmap(remoteTile));

                    Message message = mHandler.obtainMessage();

                    if (remoteTile != null && remoteTile.getBitmap() != null && remoteTile.getBitmap().length > 0) {
                        MapTile.Companion.insertTile(remoteTile, remoteTile.getBitmap());
                        message.what = TileHandler.TILE_LOADED;
                    } else {
                        message.what = TileHandler.TILE_NOT_LOADED;
                    }

                    mHandler.sendMessage(message);

                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }

    }
}

File: app/src/main/java/org/silvermoon/osm_kotlin/concurrency/RemoteAsyncTileLoader.kt
package org.silvermoon.osm_kotlin.concurrency

import android.content.Context
import android.os.Handler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import org.silvermoon.osm_kotlin.mapunits.OsmMapViewBase
import org.silvermoon.osm_kotlin.mapunits.Tile
import org.silvermoon.osm_kotlin.model.AppDatabase
import org.silvermoon.osm_kotlin.model.entities.TileEntity

class RemoteAsyncTileLoader(context: Context, private val mHandler: Handler) {

    private val scope = CoroutineScope(Dispatchers.IO + Job())
    private val mRequestsQueue = Channel<Tile>(Channel.UNLIMITED)
    private val mRequestTile = RequestTile()
    private val tileDao = AppDatabase.getDatabase(context).tileDao()

    init {
        for (i in 0 until NB_THREAD_FOR_TILE_REQUEST) {
            scope.launch {
                for (tile in mRequestsQueue) {
                    loadTileAndSendMessage(tile)
                }
            }
        }
    }

    fun queueTileRequest(tile: Tile) {
        // Make sure the zoom is not > to the actual max Tile OSM zoom
        if (tile.zoom > OsmMapViewBase.MIN_ZOOM_LEVEL_FOR_TILES)
            return

        scope.launch {
            mRequestsQueue.send(tile)
        }
    }

    fun interruptThreads() {
        // scope.cancel()
    }

    private suspend fun loadTileAndSendMessage(remoteTile: Tile?) {

        if (remoteTile != null && remoteTile.key != null) {
            try {
                remoteTile.bitmap = mRequestTile.loadBitmap(remoteTile)

                val message = mHandler.obtainMessage()

                if (remoteTile.bitmap != null && remoteTile.bitmap!!.isNotEmpty()) {
                    val entity = TileEntity(
                        row = remoteTile.mapY,
                        col = remoteTile.mapX,
                        zoom = remoteTile.zoom,
                        image = remoteTile.bitmap!!
                    )
                    tileDao.insertTile(entity)
                    message.what = TileHandler.TILE_LOADED
                } else {
                    message.what = TileHandler.TILE_NOT_LOADED
                }

                mHandler.sendMessage(message)

            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    companion object {
        private const val NB_THREAD_FOR_TILE_REQUEST = 3
    }
}
